includes/woopay/class-woopay-utilities.php

```diff
@@ -288,6 +288,41 @@ public static function encrypt_and_sign_data( $data ) {
 		];
 	}
 
+	/**
+	 * Decode encrypted and signed data and return it.
+	 *
+	 * @param array $data The session, iv, and hash data for the encryption.
+	 * @return mixed The decoded data.
+	 */
+	public static function decrypt_signed_data( $data ) {
+		$blog_token = ( self::get_woopay_url() === self::DEFAULT_WOOPAY_URL ) ? Jetpack_Options::get_option( 'blog_token' ) : 'dev_mode';
+
+		// Decode the data.
+		$decoded_data_request = array_map( 'base64_decode', $data );
+		$passphrase           = ( defined( 'WOOPAY_BLOG_ID' ) && WOOPAY_BLOG_ID === \Jetpack_Options::get_option( 'id' ) ) ? $blog_token : 'dev_mode';
+
+		if ( empty( $passphrase ) ) {
+			return null;
+		}
+
+		// Verify the HMAC hash before decryption to ensure data integrity.
+		$computed_hash = hash_hmac( 'sha256', $decoded_data_request['data'], $passphrase );
+
+		// If the hashes don't match, the message may have been tampered with.
+		if ( ! hash_equals( $computed_hash, $decoded_data_request['hash'] ) ) {
+			return null;
+		}
+
+		// Decipher the data data using the passphrase and the IV.
+		$decrypted_data = openssl_decrypt( $decoded_data_request['data'], 'aes-256-cbc', $passphrase, OPENSSL_RAW_DATA, $decoded_data_request['iv'] );
+
+		if ( false === $decrypted_data ) {
+			return null;
+		}
+
+		return $decrypted_data;
+	}
```

Comment:
It seems like `$blog_token` and `$passphrase` are redundant here. Should we make the following changes instead?

```suggestion
	public static function decrypt_signed_data( $data ) {
		$store_blog_token = ( self::get_woopay_url() === self::DEFAULT_WOOPAY_URL ) ? Jetpack_Options::get_option( 'blog_token' ) : 'dev_mode';

		// If the store blog token is empty, we can't decrypt the data.
		if ( empty( $store_blog_token ) ) {
			return null;
		}

		// Decode the data.
		$decoded_data_request = array_map( 'base64_decode', $data );

		// Verify the HMAC hash before decryption to ensure data integrity.
		$computed_hash = hash_hmac( 'sha256', $decoded_data_request['data'], $store_blog_token );

		// If the hashes don't match, the message may have been tampered with.
		if ( ! hash_equals( $computed_hash, $decoded_data_request['hash'] ) ) {
			return null;
		}

		// Decipher the data using the passphrase and the IV.
		$decrypted_data = openssl_decrypt( $decoded_data_request['data'], 'aes-256-cbc', $store_blog_token, OPENSSL_RAW_DATA, $decoded_data_request['iv'] );

		if ( false === $decrypted_data ) {
			return null;
		}

		return $decrypted_data;
	}
```

Reply:
That's true! I fixed it in [fe9c01f](https://github.com/Automattic/woocommerce-payments/pull/8849/commits/fe9c01ff7a537acb836e3fb621cf0863db9ca2f2).